// afc/components/helpers/Storage.ts
/**
 * Created by jean.rennes on 05/10/2017.
 */

namespace AFC {
    export namespace Tools {
        function _dummy(): null {
            return null;
        }
        
        function _storage(): Storage {
            let storage: Storage;
            try {
                storage = localStorage;
            } catch (e) {
                storage = {
                    length: 0,
                    clear: _dummy,
                    getItem: _dummy,
                    key: _dummy,
                    removeItem: _dummy,
                    setItem: _dummy
                };
            }
            return storage;
        }
        
        export namespace Storage {
            export function setItem(key: string, value?: any): void {
                (value !== void 0)  ? _storage().setItem(key, JSON.stringify(value)) : _storage().removeItem(key);
            }
            
            export function getItem(key: string): any {
                let value: string = _storage().getItem(key);
                if (value !== null) {
                    return JSON.parse(value);
                }
            }
            
            export function getKeys(): string[] {
                let keys: string[] = [];
                let storage: Storage = _storage();
                
                for (let i: number = 0; i < storage.length; ++i) {
                    keys.push(storage.key(i));
                }
                return keys;
            }
        }
    }
}
