// afc/Gruntfile.js
/* 
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
//const g_path = 'D:\\Mobile\\trunk\\clients\\ITME\\nomade\\node_modules\\generic\\node_modules\\afc';
//const g_path = 'D:\\mobile\\trunk\\clients\\match\\match\\node_modules\\generic\\node_modules\\afc';
//const g_path = 'D:\\mobile\\trunk\\clients\\match\\expansion\\node_modules\\generic\\node_modules\\afc';
//const g_path = 'D:\\mobile\\trunk\\clients\\hypromat\\hypromat\\node_modules\\generic\\node_modules\\afc';
//const g_path = 'D:\\mobile\\trunk\\clients\\hypromat\\hypromatMobile\\node_modules\\analyzermobile\\node_modules\\afc';
//const g_path = 'D:\\mobile\\trunk\\clients\\picard\\node_modules\\generic\\node_modules\\afc';
//const g_path = 'D:\\mobile\\trunk\\clients\\lescomptoirsdelabio\\node_modules\\generic\\node_modules\\afc';
const g_path = 'D:\\mobile\\trunk\\clients\\asterop\\sport\\node_modules\\generic\\node_modules\\afc';
//const g_path = 'D:\\mobile\\trunk\\clients\\mrbricolage\\mrbricolage\\node_modules\\generic\\node_modules\\afc';
//const g_path = 'D:\\mobile\\trunk\\clients\\fresh\\node_modules\\generic\\node_modules\\afc';
//const g_path = 'D:\\mobile\\trunk\\clients\\asterop\\ProjectAnalysis\\node_modules\\generic\\node_modules\\afc';
//const g_path = 'D:\\mobile\\trunk\\clients\\monoprix\\node_modules\\generic\\node_modules\\afc';
//const g_path = 'D:\\mobile\\trunk\\clients\\franprix\\node_modules\\generic\\node_modules\\afc';

module.exports = function (grunt) {
      var package = grunt.file.readJSON('package.json');
      
     grunt.initConfig({
        ts: {
            default : {
                tsconfig: './tsconfig.json',
                options: {
                }                
            }
        },
        replace: {
            options: {
                patterns: [{
                    match: '__AFC_VERSION__',
                    replacement: package.version
                }],
                usePrefix: false
            },
            "./ressources/afc/js/afc.js": "./ressources/afc/js/afc.js"
        },
        sync: {
            debug: {
                files: [{
                  src: ['**', '!.**/*'],
                  dest: g_path
                }],
                updateAndDelete: true,
                verbose: true // Display log messages when copying files 
            }
        }
    });
    grunt.loadNpmTasks("grunt-ts");
    grunt.loadNpmTasks('grunt-replace');    
    grunt.loadNpmTasks('grunt-sync');    
    grunt.registerTask(
            'dist',
            'Compiles all the assets and copies the files to the build directory.',
            ['ts','replace']
            );
    grunt.registerTask(
            'build',
            'Compiles all the assets and copies the files to the build directory.',
            ['dist','sync:debug']
            );
};
